lib/pure/bitops.nim

```diff
@@ -827,199 +827,143 @@ when useBuiltinsRotate:
                          {.importc: "__rorq", header: "<immintrin.h>".}
 
 func rotl[T: SomeUnsignedInt](value: T, rot: int32): T {.inline.} =
-  ## Left-rotate bits in a `SomeUnsignedInt` value.
+  ## Left-rotate bits in a `value`.
```

Comment:
running `nim doc lib/pure/bitops.nim` on osx gives:
```
runnableExamples/@mbitops_examples40.nim.c:52:31: error: conflicting types for '__builtin_rotateright64'
N_NIMCALL(unsigned long long, __builtin_rotateright64)(unsigned long long value, unsigned long long shift);
                              ^
/Users/timothee/.cache/nim/bitops_d/runnableExamples/@mbitops_examples40.nim.c:52:31: note: '__builtin_rotateright64' is a builtin with type
      'unsigned long long (unsigned long long, long long)'
```

Reply:
@rockcavera try `{.noDecl.}`

Reply:
@timotheecour What C compiler is used by the OSX CI? If it's not Clang than this is a matter of concern

Reply:
replying to 1st question: noDecl makes it work; should we add it to all declarations? let's think before changing anything

Reply:
furthermore, I got this after changing your code:
```
/Users/timothee/.cache/nim/bitops_d/runnableExamples/@mbitops_examples36.nim.c:52:31: note: '__builtin_rotateleft64' is a builtin with type
      'unsigned long long (unsigned long long, unsigned long long)'
```

```
so __builtin_rotateleft64 uses 'unsigned long long (unsigned long long, unsigned long long)'
   __builtin_rotateright64' uses 'unsigned long long (unsigned long long, long long)'
```

Reply:
I recently had a problem importing a builtin from GCC and had to add {.noDecl.} to work. This occurred only in GCC 10, as in previous versions they worked correctly.

Here, compiling with CLANG 11.0.0 and 12.0.0, on Windows, I don't get such an error. But before, using Nim integers, the CLANG compiler gave an error and asked to use proper cints and I adjusted them.

Reply:
Changing to `clonglong`, on CLANG, on Windows, gives this error:
```
C:\Users\Jose\Desktop\rot>nim c -d:danger --cc:clang -f rotbench && rotbench
Hint: used config file 'D:\nim\config\nim.cfg' [Conf]
Hint: used config file 'D:\nim\config\config.nims' [Conf]
............................
CC: no_optimize
CC: stdlib_formatfloat.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_parseutils.nim
CC: stdlib_math.nim
CC: stdlib_unicode.nim
CC: stdlib_strutils.nim
CC: stdlib_dynlib.nim
CC: stdlib_winlean.nim
CC: stdlib_times.nim
CC: stdlib_monotimes.nim
CC: ../../.nimble/pkgs/strfmt-0.10.0/strfmt.nim
CC: ../../.nimble/pkgs/nimbench-0.2.1/nimbench/private/human_readable.nim
CC: ../../.nimble/pkgs/nimbench-0.2.1/nimbench/private/timers.nim
CC: ../../.nimble/pkgs/nimbench-0.2.1/nimbench.nim
CC: rotbench.nim
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:350:31: error: conflicting types for '__builtin_rotateright64'
N_NIMCALL(unsigned long long, __builtin_rotateright64)(unsigned long long value, long long shift);
                              ^
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:350:31: note: '__builtin_rotateright64' is a builtin with type
      'unsigned long long (unsigned long long, unsigned long long)'
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:371:31: error: conflicting types for '__builtin_rotateleft64'
N_NIMCALL(unsigned long long, __builtin_rotateleft64)(unsigned long long value, long long shift);
                              ^
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:371:31: note: '__builtin_rotateleft64' is a builtin with type
      'unsigned long long (unsigned long long, unsigned long long)'
2 errors generated.
Error: execution of an external compiler program 'D:\LLVM64\bin\clang.exe -c -w -ferror-limit=3 -DWIN32_LEAN_AND_MEAN -O3   -ID:\nim\lib -IC:\Users\Jose\Desktop\rot -o C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c.o C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c' failed with exit code: 1
```
There seems to be a difference in types between the Windows version and Osx

Reply:
* btw i'm using:
```
clang -v
Apple clang version 11.0.3 (clang-1103.0.32.59)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

* can you instrument CI to temporarily print `clang -v` on osx? in future work which should do this properly (refs https://github.com/timotheecour/Nim/issues/539)

Reply:
Compiling for --cc:i386, using CLANG 8.0.0, it gives an error if it doesn't have {.noDecl.}.
```
C:\Users\Jose\Desktop\rot>nim c -d:danger --cc:clang --cpu:i386 -f rotbench && rotbench
Hint: used config file 'D:\nim\config\nim.cfg' [Conf]
Hint: used config file 'D:\nim\config\config.nims' [Conf]
............................
CC: no_optimize
CC: stdlib_formatfloat.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_parseutils.nim
CC: stdlib_math.nim
CC: stdlib_unicode.nim
CC: stdlib_strutils.nim
CC: stdlib_dynlib.nim
CC: stdlib_winlean.nim
CC: stdlib_times.nim
CC: stdlib_monotimes.nim
CC: ../../.nimble/pkgs/strfmt-0.10.0/strfmt.nim
CC: ../../.nimble/pkgs/nimbench-0.2.1/nimbench/private/human_readable.nim
CC: ../../.nimble/pkgs/nimbench-0.2.1/nimbench/private/timers.nim
CC: ../../.nimble/pkgs/nimbench-0.2.1/nimbench.nim
CC: rotbench.nim
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:332:26: error: function declared 'fastcall' here was previously declared without calling convention
N_NIMCALL(unsigned char, __builtin_rotateright8)(unsigned char value, unsigned char shift);
                         ^
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:332:26: note: previous declaration is here
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:338:27: error: function declared 'fastcall' here was previously declared without calling convention
N_NIMCALL(unsigned short, __builtin_rotateright16)(unsigned short value, unsigned short shift);
                          ^
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:338:27: note: previous declaration is here
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:344:25: error: function declared 'fastcall' here was previously declared without calling convention
N_NIMCALL(unsigned int, __builtin_rotateright32)(unsigned int value, unsigned int shift);
                        ^
C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c:344:25: note: previous declaration is here
fatal error: too many errors emitted, stopping now [-ferror-limit=]
4 errors generated.
Error: execution of an external compiler program 'D:\LLVM32\bin\clang.exe -c -w -ferror-limit=3 -DWIN32_LEAN_AND_MEAN -O3   -ID:\nim\lib -IC:\Users\Jose\Desktop\rot -o C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c.o C:\Users\Jose\nimcache\rotbench_r\@mrotbench.nim.c' failed with exit code: 1
```


However, with the 64-bit compiler it no longer gives this error and neither in versions 11.0.0 and 12.0.0.

You may want to add {.noDecl.}.

Reply:
some more data: https://github.com/llvm-mirror/clang/blob/master/include/clang/Basic/Builtins.def shows:

BUILTIN(__builtin_rotateright64, "UWiUWiUWi", "nc")
BUILTIN(__builtin_rotateleft64, "UWiUWiUWi", "nc")
which according to definitions in https://github.com/llvm-mirror/clang/blob/master/include/clang/Basic/Builtins.def would translate to 2nd param being unsigned 64 bit value


so i don't know why clang complains with:
```
 '__builtin_rotateright64' is a builtin with type
      'unsigned long long (unsigned long long, long long)'
```

Reply:
@rockcavera @awr1
ok this explains everything: https://github.com/llvm-mirror/clang/commit/892de415b7fde609dafc4e6c1643b7eaa0150a4d

> Change signature of __builtin_rotateright64 back to unsigned

so the correct thing is unsigned, and we should use noDecl

Reply:
done!